### PR TITLE
[common]: add common chart

### PR DESCRIPTION
**Pull request by janlauber**
**What this PR does**:
- add merged `common` and `library` chart to inventory

**Notes for Reviewer**:  
The `common` charts folder is just copied from the internal gitlab.  
All `*.tpl` files from the library chart are copied to the `helpers` folder.  
Also the `README` got improved. 

**Checklist**:

- [x] Pull Request title in format `[chart]: Changed Something`
- [x] Updated documentation in the  `README.md.gotmpl` file and executed `helm-docs` 
- [x] Chart Version bumped
- [x] All commits are signed-off

### CODE REVIEW

**Comment by janfuhrer** [comment]
remove all internal links (wikis, registry, etc.) and comments that should not be published
perhaps the internal labels should also be reconsidered

File: charts/common/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+apiVersion: v2
+name: common
+description: A common Helm chart to use for creating other Helm charts
```
Comment (janfuhrer) [inline charts/common/Chart.yaml:3]: - I would add "Bedag" in the description

have a look at https://github.com/bedag/helm-charts/blob/master/charts/manifests/Chart.yaml and add
- annotations
- sources
- icon
- remove "home"

File: charts/common/.helmignore [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+*~
+# Various IDEs
+.project
+.idea/
+*.tmproj
+.vscode/
+# helm files
+.helmignore
+# gitlab files
+.gitlab-ci.yml
```
Comment (janfuhrer) [inline charts/common/.helmignore:26]: add README.md.gotmpl

File: charts/common/mkdocs.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+site_name: Changelog
+pages:
+  - Changelog: index.md
+theme: readthedocs
+site_dir: public
+markdown_extensions:
+  - toc:
+      permalink: True
```
Comment (janfuhrer) [inline charts/common/mkdocs.yml:8]: this file can be deleted

File: charts/common/templates/helpers/_labels.tpl [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+{{- /*
+library.labels.standard prints the standard Helm labels.
+The standard labels are frequently used in metadata.
+
+Based on https://docs.helm.sh/chart_best_practices/#standard-labels
+*/ -}}
+{{- define "library.labels.standard" -}}
+app.kubernetes.io/name: {{ template "library.name" . }}
+helm.sh/chart: {{ template "library.chartrefshort" . }}
+helm.nsm.bedag.ch/chartname: {{ .Chart.Name | quote }}
```
Comment (janfuhrer) [inline charts/common/templates/helpers/_labels.tpl:23]: I would change all labels with `nsm.bedag.ch` to something different (we can discuss this)

File: charts/common/values.test.yaml [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+      deploy: true
+    # end common.networkpolicy
+    controller:
+      deploy: true
+      type: "Deployment"
+
+      containers:
+        container-1:
+          image:
+            repository: registry-group.mgmtbi.ch/docker_base/nginx
```
Comment (janfuhrer) [inline charts/common/values.test.yaml:46]: remove internal registry (multiple times in this file)

File: charts/common/values.yaml [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+  # deploy has to be set to true for rendering to be applied
+  deploy: true
+# end common.networkpolicy
+
+# start common.ingress
+# ingress is used for access from outside of kubernetes
+ingress:
+  # deploy has to be set to true for rendering to be applied
+  deploy: false
+  # IngressClassName, defines the class of the ingress controller. For Bedag classes see https://confluence.se.bedag.ch/display/KUBERNETES/Service+Katalog
```
Comment (janfuhrer) [inline charts/common/values.yaml:44]: remove internal links (multiple times in this file)

File: charts/common/values.yaml [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+      # Has to be base64 encoded and should be encrypted in the ejson vault
+      # Add Variable to your CI/CD Settings "SKIP_DECRYPT" with value "" that it doesnt decrypt the cert and fails. bsp. (https://git.mgmtbi.ch/kubernetes/applications/a51_capitastra_bbv/-/jobs/2055510#L70)
+      cert: ""
+      # The key must not have a passphrase
+      key: ""
+# end common.ingress
+
+# start common.servicemonitor
+# servicemonitor is from CoreOS and lets you scrape metrics with the prometheus-operators
+# Hint: The output of the metrics endpoint must be correctly formatted, see https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md
```
Comment (janfuhrer) [inline charts/common/values.yaml:106]: update link to https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md

File: charts/common/values.yaml [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+      #  - "foo.com"
+      # validityDuration defines how long the certificate is valid (in days)
+      validityDuration: 365
+    existing:
+      # name of an existing secret with tls.crt & tls.key content
+      secret: ""
+    provided:
+      # If SSL is terminated on ingress and you have a generated (preferrably CERT-001) certificate/key
+      # Has to be base64 encoded and should be encrypted in the ejson vault
+      # Add Variable to your CI/CD Settings "SKIP_DECRYPT" with value "" that it doesnt decrypt the cert and fails. bsp. (https://git.mgmtbi.ch/kubernetes/applications/a51_capitastra_bbv/-/jobs/2055510#L70)
```
Comment (janfuhrer) [inline charts/common/values.yaml:98]: remove internal link

File: charts/common/values.yaml [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+        # foo: bar
+        # prometheus.io/path: /metrics
+        # prometheus.io/port: "8080"
+        # prometheus.io/scrape: "true"
+
+      # Extra labels on all Pods
+      # We use this for example to set the Loki logging tenant.
+      extraLabels: {}
+        # foo: bar
+        # tenant.nsm.bedag.ch/xOrgId: MyTenant
```
Comment (janfuhrer) [inline charts/common/values.yaml:246]: remove this label

File: charts/common/values.yaml [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+        # type of deployment. Can be "Recreate" or "RollingUpdate". Comment out for using default ("Recreate")
+        #type: "RollingUpdate"
+        # maxSurge is only needed when type is "RollingUpdate". The maximum number of pods that can be scheduled above the desired number of pods. Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%). Comment out for using default (1)
+        #maxSurge: 2
+        # maxUnavailable is only needed when type is "RollingUpdate". The maximum number of pods that can be unavailable during the update. Value can be an absolute number (ex: 5) or a percentage of desired pods (ex:10%). Comment out for using default (1)
+        #maxUnavailable: 50%
+      # END ONLY FOR DEPLOYMENTS
+
+      # START ONLY FOR JOBS OR CRONJOBS
+      # see https://v1-23.docs.kubernetes.io/docs/reference/kubernetes-api/workload-resources/job-v1/ for more details
```
Comment (janfuhrer) [inline charts/common/values.yaml:304]: change to general link: https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/job-v1/

File: charts/common/values.yaml [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+      # Priority indicates the importance of a Pod relative to other Pods. If a Pod cannot be scheduled,
+      # the scheduler tries to preempt (evict) lower priority Pods to make scheduling of the pending Pod possible.
+      #priorityClassName: bronze-tier-priority
+
+      # Restart policy for all containers within the pod. One of Always, OnFailure, Never. Default to Always.
+      #restartPolicy: "Never"
+
+      # podSecurityContext holds pod-level security attributes and common container settings.
+      # Some fields are also present in container.securityContext. Field values of containerSecurityContext take precedence over field values of podSecurityContext.
+      # ref: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podsecuritycontext-v1-core
```
Comment (janfuhrer) [inline charts/common/values.yaml:363]: change to general link: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

File: charts/common/values.yaml [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+
+          # command Entrypoint array. Not executed within a shell. The docker image's ENTRYPOINT is used if this is not provided.
+          #command: []
+
+          # args Arguments to the entrypoint. The docker image's CMD is used if this is not provided.
+          #args: []
+
+          # containerSecurityContext holds security configuration that will be applied to a container.
+          # Some fields are present in both containerSecurityContext and podSecurityContext. When both are set, the values in containerSecurityContext take precedence.
+          # ref: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#securitycontext-v1-core
```
Comment (janfuhrer) [inline charts/common/values.yaml:417]: change to general link: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

File: charts/common/values.test.yaml [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+      completionMode: "Indexed"
+      backoffLimit: 2
+      activeDeadlineSeconds: 120
+      ttlSecondsAfterFinished: 0
+      automountServiceAccountToken: false
+      restartPolicy: "Never"
+
+      # podSecurityContext holds pod-level security attributes and common container settings.
+      # Some fields are also present in container.securityContext. Field values of containerSecurityContext take precedence over field values of podSecurityContext.
+      # ref: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podsecuritycontext-v1-core
```
Comment (janfuhrer) [inline charts/common/values.test.yaml:203]: update to general link: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

File: charts/common/templates/helpers/_labels.tpl [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+{{- range $k, $v := . }}
+{{ $k }}: {{ $v | quote }}
+{{- end -}}
+{{- end -}}
+
+{{- /*
+library.labels.standard prints the standard Helm labels.
+The standard labels are frequently used in metadata.
+
+Based on https://docs.helm.sh/chart_best_practices/#standard-labels
```
Comment (janfuhrer) [inline charts/common/templates/helpers/_labels.tpl:18]: update link to https://helm.sh/docs/chart_best_practices/#standard-labels

File: charts/common/values.yaml [hunk 15]
```diff
@@ -0,0 +1,600 @@
...
+#fullnameOverride: ""
+
+# defaultTag is used when no imageTag is specified in .Values.components.*.controller.image.tag
+defaultTag: latest
+
+# timezone to set as environment variable 'TZ' in each pod. Comment out for using default ("Europe/Zurich")
+#timezone: "Europe/Zurich"
+
+# secrets contains the pullSecret for the container registry to pull images.
+# In a Bedag namespace this value does not have to be set for Dockerhub or the Bedag registry
```
Comment (janfuhrer) [inline charts/common/values.yaml:26]: remove this comment

File: charts/common/values.yaml [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+      cert: ""
+      # The key must not have a passphrase
+      key: ""
+# end common.ingress
+
+# start common.servicemonitor
+# servicemonitor is from CoreOS and lets you scrape metrics with the prometheus-operators
+# Hint: The output of the metrics endpoint must be correctly formatted, see https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md
+# - /actuator/metrics in Spring Boot based applications from Bedag, this endpoint is not served (plain json output)
+# - /metrics in Spring Boot based applications from Bedag this is the correct endpoint to scrape metrics
```
Comment (janfuhrer) [inline charts/common/values.yaml:108]: remove "Bedag" in this comments

**Review by janfuhrer — CHANGES_REQUESTED** [review]

File: charts/common/mkdocs.yml (hunk 3 above)
Comment (janlauber) [reply charts/common/mkdocs.yml:8]: done

File: charts/common/values.test.yaml (hunk 5 above)
Comment (janlauber) [reply charts/common/values.test.yaml]: done

File: charts/common/values.yaml (hunk 6 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/values.yaml (hunk 7 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/values.yaml (hunk 10 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/values.yaml (hunk 11 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/values.yaml (hunk 12 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/values.test.yaml (hunk 13 above)
Comment (janlauber) [reply charts/common/values.test.yaml]: done

File: charts/common/templates/helpers/_labels.tpl (hunk 14 above)
Comment (janlauber) [reply charts/common/templates/helpers/_labels.tpl]: done

File: charts/common/values.yaml (hunk 15 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/values.yaml (hunk 16 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/values.yaml [hunk 17]
```diff
@@ -0,0 +1,597 @@
...
+      # Add Variable to your CI/CD Settings "SKIP_DECRYPT" with value "" that it doesnt decrypt the cert and fails.
+      cert: ""
+      # The key must not have a passphrase
+      key: ""
+# end common.ingress
+
+# start common.servicemonitor
+# servicemonitor is from CoreOS and lets you scrape metrics with the prometheus-operators
+# Hint: The output of the metrics endpoint must be correctly formatted, see https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md
+# - /actuator/metrics in Spring Boot based applications from Bedag, this endpoint is not served (plain json output)
```
Comment (janfuhrer) [inline charts/common/values.yaml:106]: remove "Bedag" as well

File: charts/common/values.yaml [hunk 18]
```diff
@@ -0,0 +1,597 @@
...
+  deploy: false
+  # basicAuth is a dictionary for defining values for setting up basic Authentication
+  basicAuth:
+    # enabled when set to 'true', adds basic authentication to all endpoints
+    enabled: false
+    # existingSecret if not empty (""), points to an existing secret (matching name of the resource)
+    existingSecret: ""
+    # newSecret is a dictionary for defining key/value pairs to be stored in a new secret
+    newSecret: {}
+      #username: bedag
```
Comment (janfuhrer) [inline charts/common/values.yaml:119]: I would change `username` and `custom-username` to something other than bedag

File: charts/common/Chart.yaml [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+name: common
+description: "Bedag's common Helm chart to use for creating other Helm charts"
+version: 7.0.0
+type: application
+icon: "https://www.bedag.ch/wGlobal/wGlobal/layout/images/logo.svg"
+keywords:
+- Bedag
+- Common
+sources:
+- https://github.com/bedag/helm-charts/tree/master/charts/manifests
```
[the remark this answers is not in the document]
Comment (janlauber) [reply charts/common/Chart.yaml]: done

File: charts/common/Chart.yaml [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+maintainers:
+  - email: sre@bedag.ch
+    name: Site Reliability Engineering
+annotations:
+  artifacthub.io/containsSecurityUpdates: "false"
+  artifacthub.io/prerelease: "false"
+  artifacthub.io/license: Apache-2.0
+  artifacthub.io/changes: |
+    - "[Added]: Functionality for environment from source"
```
[the remark this answers is not in the document]
Comment (janlauber) [reply charts/common/Chart.yaml]: done

File: charts/common/Chart.yaml [hunk 21]
```diff
@@ -0,0 +1,20 @@
+apiVersion: v2
+name: common
+description: "Bedag's common Helm chart to use for creating other Helm charts"
+version: 7.0.0
+type: application
```
[the remark this answers is not in the document]
Comment (janlauber) [reply charts/common/Chart.yaml:15]: done

File: charts/common/values.test.yaml [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+        runAsGroup: 1001
+        runAsNonRoot: true
+        runAsUser: 1001
+
+      # start containers
+      # dictionary of containers in a pod
+      containers:
+        job:
+          image:
+            repository: registry-group.mgmtbi.ch/perl
```
[the remark this answers is not in the document]
Comment (janlauber) [reply charts/common/values.test.yaml]: done

File: charts/common/values.test.yaml [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+        runAsGroup: 1001
+        runAsNonRoot: true
+        runAsUser: 1001
+
+      # start containers
+      # dictionary of containers in a pod
+      containers:
+        job:
+          image:
+            repository: registry-group.mgmtbi.ch/busybox
```
[the remark this answers is not in the document]
Comment (janlauber) [reply charts/common/values.test.yaml]: done

File: charts/common/values.test.yaml [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+# end common.ingress
+
+# start common.servicemonitor
+servicemonitor:
+  deploy: true
+  basicAuth:
+    enabled: true
+    existingSecret: ""
+    newSecret:
+      username: bedag
```
[the remark this answers is not in the document]
Comment (janlauber) [reply charts/common/values.test.yaml]: done

File: charts/common/values.yaml (hunk 17 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/values.yaml (hunk 18 above)
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/templates/_container.yaml [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+  env:
+    - name: TZ
+      value: {{ default "Europe/Zurich" $root.Values.timezone | quote }}
+  {{- with $containerValues.env  }}
+    {{- if (kindIs "slice" .) }}
+      {{- toYaml . | nindent 4 }}
+    {{- else }}
+      {{- range $key, $value := . }}
+    - name: {{ $key }}
+      value: {{ $value }}
```
Comment (janfuhrer) [inline charts/common/templates/_container.yaml:25]: change this line to `value: {{ $value | quote }}`
At the moment there is a small bug with numbers in env maps, so we can fix it directly with this new release

File: charts/common/templates/_networkpolicy.yaml [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+      app.kubernetes.io/name: {{ template "library.name" $root }}
+      app.kubernetes.io/component: {{ $name }}
+  policyTypes:
+    - Ingress
+  ingress:
+    - from:
+      - namespaceSelector:
+          matchLabels:
+            # trusted label is used by Bedag on platformspecific services (e.g. Ingress Controller) for them to access the selected pods
+            trusted: {{ default "false" ($networkpolicy.trusted | quote) }}
```
Comment (janfuhrer) [inline charts/common/templates/_networkpolicy.yaml:43]: @oliverbaehler do we still need the "trusted" function?

File: charts/common/values.yaml [hunk 27]
```diff
@@ -0,0 +1,596 @@
...
+        #  label2.2: "label2"
+        # sessionAffinity supports "ClientIP" and "None". Used to maintain session affinity. Default value is "None".
+        #sessionAffinity: "ClientIP"
+    # end common.service
+    # start common.networkpolicy
+    # networkpolicy allows access from outside of the namespace or from other pods
+    networkpolicy:
+      # deploy has to be set to true for rendering to be applied
+      deploy: false
+      # trusted is optional with a default value of "false"
```
Comment (janfuhrer) [inline charts/common/values.yaml:213]: if we keep the trusted function, add this comment to clarify the effects
`, it allows traffic from namespaces with label "trusted" (e.g. for system namespaces like ingress-controller)`

**Review by janfuhrer — CHANGES_REQUESTED** [review]

File: charts/common/values.yaml [hunk 28]
```diff
@@ -0,0 +1,596 @@
...
+      cert: ""
+      # The key must not have a passphrase
+      key: ""
+# end common.ingress
+
+# start common.servicemonitor
+# servicemonitor is from CoreOS and lets you scrape metrics with the prometheus-operators
+# Hint: The output of the metrics endpoint must be correctly formatted, see https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md
+# - /actuator/metrics in Spring Boot based applications, this endpoint is not served (plain json output)
+# - /metrics in Spring Boot based applications, this is the correct endpoint to scrape metrics
```
[the remark this answers is not in the document]
Comment (janlauber) [reply charts/common/values.yaml]: done

File: charts/common/templates/_container.yaml (hunk 25 above)
Comment (janlauber) [reply charts/common/templates/_container.yaml]: done

**Review by oliverbaehler — CHANGES_REQUESTED** [review]
See comments

File: charts/common/values.test.yaml [hunk 29]
```diff
@@ -0,0 +1,307 @@
+# start common.ingress
```
[the remark this answers is not in the document]
Comment (janlauber) [reply charts/common/values.test.yaml]: done

File: charts/common/templates/_networkpolicy.yaml (hunk 26 above)
Comment (oliverbaehler) [reply charts/common/templates/_networkpolicy.yaml:43]: @janlauber Just remove that trusted field, to keep it simple.

Comment (janlauber) [reply charts/common/templates/_networkpolicy.yaml]: @oliverbaehler I've removed the whole array obj `namespaceSelector`.
Also removed the comment in the `values.yaml` file

File: .github/workflows/release.yml [hunk 30]
```diff
@@ -6,10 +6,6 @@ on:
   push:
     branches:
       - master
-
-permissions:
-  contents: read
-
```
Comment (oliverbaehler) [inline .github/workflows/release.yml:12]: @janlauber mach nid so

**Review by oliverbaehler — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml (hunk 30 above)
Comment (janlauber) [reply .github/workflows/release.yml]: @oliverbaehler okey

**Review by oliverbaehler — APPROVED** [review]
lgtm

**Review by janfuhrer — APPROVED** [review]
lgtm

**Pull request merged**